common/lib/xmodule/xmodule/html_module.py

```diff
@@ -117,9 +120,18 @@ def get_html(self):
         """ Returns html required for rendering the block. """
         if self.data:
             data = self.data
-            if getattr(self.runtime, 'anonymous_student_id', None):
-                data = data.replace("%%USER_ID%%", self.runtime.anonymous_student_id)
-            data = data.replace("%%COURSE_ID%%", str(self.scope_ids.usage_id.context_key))
+            course_id = str(self.scope_ids.usage_id.context_key)
+            # Have to catch NoSuchServiceError in case we are wrapped by a ConditionalBlock
```

Comment:
It's only happening with one test, the [`test_conditional_module`](https://github.com/edx/edx-platform/blob/da09bcadc571b92a42d3cb9331454f624fec0977/common/lib/xmodule/xmodule/tests/test_conditional.py#L244), so the fix may be in the runtime setup for this test? I've tried a bunch of things but failed to locate where the user service needs to be injected.

Exception occurs at [this line](https://github.com/edx/edx-platform/blob/da09bcadc571b92a42d3cb9331454f624fec0977/common/lib/xmodule/xmodule/tests/test_conditional.py#L294), and traceback is pasted below, with all the stdout and log output.

6d0575e4a279e1ab0e4278a2e79d0e430b75fd63 removes this try/catch, so we'll also see this fail in the python tests.

Any ideas?

```
common/lib/xmodule/xmodule/tests/test_conditional.py:306:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
common/lib/xmodule/xmodule/conditional_module.py:291: in handle_ajax
    fragments = [child.render(STUDENT_VIEW).to_dict() for child in self.get_display_items()]
common/lib/xmodule/xmodule/conditional_module.py:291: in <listcomp>
    fragments = [child.render(STUDENT_VIEW).to_dict() for child in self.get_display_items()]
/edx/src/XBlock/xblock/core.py:198: in render
    return self.runtime.render(self, view, context)
common/lib/xmodule/xmodule/x_module.py:2042: in render
    return self.__getattr__('render')(block, view_name, context)
common/lib/xmodule/xmodule/x_module.py:1427: in render
    return super().render(block, view_name, context=context)
/edx/src/XBlock/xblock/runtime.py:847: in render
    frag = view_fn(context)
common/lib/xmodule/xmodule/html_module.py:94: in student_view
    fragment = Fragment(self.get_html())
lms/djangoapps/edxnotes/decorators.py:31: in get_html
    return original_get_html(self, *args, **kwargs)
common/lib/xmodule/xmodule/html_module.py:123: in get_html
    user_service = self.runtime.service(self, 'user')
common/lib/xmodule/xmodule/x_module.py:2074: in service
    service = self._descriptor_system.service(block, service_name)
common/lib/xmodule/xmodule/x_module.py:1620: in service
    service = super().service(block=block, service_name=service_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <xmodule.modulestore.xml.ImportSystem object at 0x7f0163591a30>
block = <HtmlBlockWithMixins @8F20 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'co...rce_code=None, use_latex_compiler=False, xml_attributes={'filename': ['html/secret_page.xml', 'html/secret_page.xml']}>
service_name = 'user'
    def service(self, block, service_name):
        """Return a service, or None.

        Services are objects implementing arbitrary other interfaces.  They are
        requested by agreed-upon names, see [XXX TODO] for a list of possible
        services.  The object returned depends on the service requested.

        XBlocks must announce their intention to request services with the
        `XBlock.needs` or `XBlock.wants` decorators.  Use `needs` if you assume
        that the service is available, or `wants` if your code is flexible and
        can accept a None from this method.

        Runtimes can override this method if they have different techniques for
        finding and delivering services.

        Arguments:
            block (XBlock): this block's class will be examined for service
                decorators.
            service_name (str): the name of the service requested.

        Returns:
            An object implementing the requested service, or None.

        """
        declaration = block.service_declaration(service_name)
        if declaration is None:
            raise NoSuchServiceError(f"Service {service_name!r} was not requested.")
        service = self._services.get(service_name)
        if service is None and declaration == "need":
>           raise NoSuchServiceError(f"Service {service_name!r} is not available.")
E           xblock.exceptions.NoSuchServiceError: Service 'user' is not available.

/edx/src/XBlock/xblock/runtime.py:1125: NoSuchServiceError
--------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------
Starting import
Importing conditional_and_poll
Course:  <CourseBlockWithMixins @B591 license=None, parent=None, name=None, tags=[], display_name='Justice', children=[BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'chapter', 'Staff'), BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'chapter', 'Problems_with_Condition')], default_time_limit_minutes=None, due=None, exam_review_rules='', hide_after_due=False, is_entrance_exam=False, is_onboarding_exam=False, is_practice_exam=False, is_proctored_enabled=False, is_time_limited=False, position=None, relative_weeks_due=None, xml_attributes={'filename': ['course/2013_Spring.xml', 'course/2013_Spring.xml']}, advanced_modules=[], advertised_start='Spring 2013', allow_anonymous=True, allow_anonymous_to_peers=False, allow_proctoring_opt_out=False, allow_public_wiki_access=False, allow_unsupported_xblocks=False, announcement=None, banner_image='images_course_image.jpg', catalog_visibility='both', ccx_connector='', cert_html_view_enabled=True, cert_html_view_overrides={}, cert_name_long='', cert_name_short='', certificate_available_date=None, certificates={}, certificates_display_behavior='end', certificates_show_before_end=False, cohort_config={}, cosmetic_display_price=0, course_edit_method='Studio', course_image='images_course_image.jpg', course_survey_name=None, course_survey_required=False, course_visibility='private', create_zendesk_tickets=True, css_class='', disable_progress_graph=False, discussion_blackouts=[], discussion_link=None, discussion_sort_alpha=False, discussion_topics={'General': {'id': 'i4x-HarvardX-ER22x-course-2013_Spring'}}, display_coursenumber='', display_organization=None, due_date_display_format=None, enable_ccx=False, enable_proctored_exams=False, enable_subsection_gating=False, enable_timed_exams=False, end=None, end_of_course_survey_url=None, enrollment_domain=None, enrollment_end=None, enrollment_start=None, entrance_exam_enabled=False, entrance_exam_id=None, entrance_exam_minimum_score_pct=65, grading_policy={'GRADER': [{'type': 'Homework', 'short_label': 'HW', 'min_count': 12, 'drop_count': 2, 'weight': 0.15}, {'type': 'Lab', 'min_count': 12, 'drop_count': 2, 'weight': 0.15}, {'type': 'Midterm Exam', 'short_label': 'Midterm', 'min_count': 1, 'drop_count': 0, 'weight': 0.3}, {'type': 'Final Exam', 'short_label': 'Final', 'min_count': 1, 'drop_count': 0, 'weight': 0.4}], 'GRADE_CUTOFFS': {'Pass': 0.5}}, hide_progress_tab=None, highlights_enabled_for_messaging=False, html_textbooks=[], info_sidebar_name='Course Handouts', instructor_info={'instructors': []}, invitation_only=False, is_new=None, issue_badges=True, language=None, learning_info=[], lti_passports=[], max_student_enrollments_allowed=None, minimum_grade_credit=0.8, mobile_available=True, no_grade=False, other_course_settings={}, pdf_textbooks=[], pre_requisite_courses=[], proctoring_escalation_email=None, proctoring_provider='null', remote_gradebook={}, self_paced=False, show_calculator=False, show_timezone=True, social_sharing_url=None, start=datetime.datetime(2099, 1, 1, 0, 0, tzinfo=<UTC>), tabs=[<lms.djangoapps.courseware.tabs.CourseInfoTab object at 0x7f01634a4c70>, <lms.djangoapps.courseware.tabs.CoursewareTab object at 0x7f01634a44f0>, <lms.djangoapps.courseware.tabs.TextbookTabs object at 0x7f01634a4430>, <lms.djangoapps.discussion.plugins.DiscussionTab object at 0x7f01634a4460>, <lms.djangoapps.course_wiki.tab.WikiTab object at 0x7f016350ea90>, <lms.djangoapps.courseware.tabs.ProgressTab object at 0x7f016350e970>], teams_configuration=<TeamsConfig default_max_team_size=50 teamsets=[]>, textbooks=[], video_thumbnail_image='images_course_image.jpg', video_upload_pipeline={}, wiki_slug='ER22x'>
id:  HarvardX/ER22x/2013_Spring
module:  <ConditionalBlockWithMixins @B2C1 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'sequential', 'sequential_673cfeb23b2f'), name=None, tags=[], display_name='Conditional', children=[BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'html', 'secret_page')], conditional_attr='attempted', conditional_message='', conditional_value='True', due=None, hide_after_due=False, is_entrance_exam=False, position=None, relative_weeks_due=None, show_tag_list=[], sources_list=[BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'problem', 'choiceprob')], xml_attributes={'attempted': 'True', 'sources': 'i4x://HarvardX/ER22x/problem/choiceprob', 'filename': ['conditional/condone.xml', 'conditional/condone.xml']}>
module children:  [<HtmlBlockWithMixins @8F20 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'conditional', 'condone'), name=None, tags=[], display_name='Secret Page', data='<p>This is a secret!</p>', editor='visual', source_code=None, use_latex_compiler=False, xml_attributes={'filename': ['html/secret_page.xml', 'html/secret_page.xml']}>]
module display items (children):  [<HtmlBlockWithMixins @8F20 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'conditional', 'condone'), name=None, tags=[], display_name='Secret Page', data='<p>This is a secret!</p>', editor='visual', source_code=None, use_latex_compiler=False, xml_attributes={'filename': ['html/secret_page.xml', 'html/secret_page.xml']}>]
html type:  <class 'str'>
html:  (('conditional_ajax.html',
  {'ajax_url': 'i4x://HarvardX/ER22x/conditional/condone/xmodule_handler',
   'depends': 'i4x-HarvardX-ER22x-problem-choiceprob',
   'element_id': 'i4x-HarvardX-ER22x-conditional-condone'}),
 {})
gdi= [<HtmlBlockWithMixins @8F20 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'conditional', 'condone'), name=None, tags=[], display_name='Secret Page', data='<p>This is a secret!</p>', editor='visual', source_code=None, use_latex_compiler=False, xml_attributes={'filename': ['html/secret_page.xml', 'html/secret_page.xml']}>]
ajax:  {'fragments': [{'content': "(('conditional_module.html',\n  {'message': '',\n   'module': <ConditionalBlockWithMixins @B2C1 parent=BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'sequential', 'sequential_673cfeb23b2f'), name=None, tags=[], display_name='Conditional', children=[BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'html', 'secret_page')], conditional_attr='attempted', conditional_message='', conditional_value='True', due=None, hide_after_due=False, is_entrance_exam=False, position=None, relative_weeks_due=None, show_tag_list=[], sources_list=[BlockUsageLocator(CourseLocator('HarvardX', 'ER22x', '2013_Spring', None, None), 'problem', 'choiceprob')], xml_attributes={'attempted': 'True', 'sources': 'i4x://HarvardX/ER22x/problem/choiceprob', 'filename': ['conditional/condone.xml', 'conditional/condone.xml']}>}),\n {})"}], 'message': False}
---------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------
ERROR    xmodule.seq_module:seq_module.py:121 Unable to load child when parsing Sequence. Continuing...
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 118, in definition_from_xml
    child_block = system.process_xml(etree.tostring(child, encoding='unicode'))
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 213, in process_xml
    child.parent = descriptor.location
AttributeError: 'VerticalBlockWithMixins' object has no attribute 'location'
ERROR    xmodule.seq_module:seq_module.py:121 Unable to load child when parsing Sequence. Continuing...
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 118, in definition_from_xml
    child_block = system.process_xml(etree.tostring(child, encoding='unicode'))
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 213, in process_xml
    child.parent = descriptor.location
AttributeError: 'WrapperBlockWithMixins' object has no attribute 'location'
```

Reply:
Hmm, according to the traceback, the `service()` methods finds the `needs` declaration but it is unable to find the service in the `runtime._services` dict. The [TestModuleSystem](https://github.com/edx/edx-platform/pull/28440/files#diff-caad0b6dd6279f98a4b30df46f5bbd1b895e5167e188658de1d9ae308c06c560L127) is the `block.runtime` here right? Is the user service present right after `get_test_system()` in `test_conditional_module()`?

Reply:
>  Is the user service present right after `get_test_system(`) in `test_conditional_module()`?

Yes, the `TestModuleSystem` returned by `get_test_system()` [provides a user service](https://github.com/open-craft/edx-platform/blob/jill/BD-13-user-service/common/lib/xmodule/xmodule/tests/__init__.py#L144).

But the issue is that the child blocks for the ConditionalXBlock don't have a `_module_system` set at all, because the stack trace shows that it's [looking for the user service in the descriptor](https://github.com/open-craft/edx-platform/blob/jill/BD-13-user-service/common/lib/xmodule/xmodule/x_module.py#L2074):

```
 common/lib/xmodule/xmodule/x_module.py:2074: in service
    service = self._descriptor_system.service(block, service_name)
```

The ConditionalXBlock has a user service, but its children don't. I can fix this by passing in a user_service to the [XMLModuleStore constructor](https://github.com/open-craft/edx-platform/blob/jill/BD-13-user-service/common/lib/xmodule/xmodule/tests/test_conditional.py#L236), but I'm not sure if this fixes the issue, or just fixes the tests.

What is the XMLModuleStore used for? Is it just for importing XML, but not rendering it?

Reply:
@symbolist I sorted it out -- the issue was just an artifact of how the descriptor was being initialized in the ConditionalBlockXmlTest test.

What the [`inner_get_module`](https://github.com/edx/edx-platform/blob/da09bcadc571b92a42d3cb9331454f624fec0977/common/lib/xmodule/xmodule/tests/test_conditional.py#L253) mock missing was the [`descriptor.bind_for_student`](https://github.com/edx/edx-platform/blob/da09bcadc571b92a42d3cb9331454f624fec0977/common/lib/xmodule/xmodule/tests/__init__.py#L123) call, but once I worked that out, it became clear that all the [recursive and messy mocking of `get_module`](https://github.com/edx/edx-platform/blob/da09bcadc571b92a42d3cb9331454f624fec0977/common/lib/xmodule/xmodule/tests/test_conditional.py#L253-L272) done for this test could be replaced with a much simpler [`get_module_for_location`](https://github.com/edx/edx-platform/pull/28440/commits/070e593691f6173c5e275258acbb077aea7bda3c#diff-390fa336e0070d8ce05adc1742f05e93d8ea85569d13d2b58bdcda120d73fdd6R229-R231) method.

cf 070e593

Reply:
Nice!